reco_utils/recommender/tfidf/tfidf_utils.py

```diff
@@ -0,0 +1,400 @@
+            
+            # When defining a custome tokenizer with TfidfVectorizer, the tokenization is applied in the fit function
+            tf = TfidfVectorizer(tokenizer=tokenize, analyzer='word', ngram_range=ngram_range, min_df=min_df, stop_words='english')
+            vectors_tokenized = vectors
+
+        elif self.tokenization_method == 'none':
+            # No tokenization applied
+            tf = TfidfVectorizer(analyzer='word', ngram_range=ngram_range, min_df=min_df, stop_words='english')
+            vectors_tokenized = vectors
+        
+        # Save to class variable
+        self.tf = tf
+
+        return tf, vectors_tokenized
+    
+    def fit_tfidf(self, tf, vectors_tokenized):
+        """ Fit TF-IDF vectorizer to the cleaned and tokenized text.
+
+        Args:
+            tf (TfidfVectorizer): Scikit-learn TfidfVectorizer object defined in .tokenize_text().
+            vectors_tokenized (pd.Series): Each row contains tokens for respective documents separated by spaces.
+        
+        Returns:
+            tfidf_matrix (scipy.sparse.csr.csr_matrix): TF-IDF matrix.
+        """
+        
+        # Create TF-IDF matrix
+        self.tfidf_matrix = tf.fit_transform(vectors_tokenized)
+
+        return self.tfidf_matrix
+    
+    def get_tokens(self):
+        """ Return the tokens generated by the TF-IDF vectorizer.
+
+        Returns:
+            self.tokens (dict): Dictionary of tokens generated by the TF-IDF vectorizer.
+        """
+        try:
+            self.tokens = self.tf.vocabulary_
+        except:
+            self.tokens = 'Run .tokenize_text() and .fit_tfidf() first'
+        return self.tokens
+    
+    def get_stop_words(self):
+        """ Return the stop words excluded in the TF-IDF vectorizer.
+
+        Returns:
+            self.stop_words (frozenset): Frozenset of stop words used by the TF-IDF vectorizer (can be converted to list).
+        """
+        try:
+            self.stop_words = self.tf.get_stop_words()
+        except:
+            self.stop_words = 'Run .tokenize_text() and .fit_tfidf() first'
+        return self.stop_words
+    
+    def __create_full_recommendation_dictionary(self, df_clean):
+        """ Create the full recommendation dictionary containing all recommendations for all items.
+        
+        Args:
+            df_clean (pd.DataFrame): Dataframe with cleaned text.
+        """
+
+        # Similarity measure
+        cosine_sim = linear_kernel(self.tfidf_matrix, self.tfidf_matrix)
+
+        results = {}
+        for idx, row in df_clean.iterrows():
+            similar_indices = cosine_sim[idx].argsort()[:-(len(df_clean)+1):-1]
+            similar_items = [(cosine_sim[idx][i], df_clean[self.id_col][i], df_clean[self.title_col][i]) for i in similar_indices]
+            results[row[self.id_col]] = similar_items[1:]
+        
+        # Save to class
+        self.recommendations = results
+    
+    def __organize_results_as_tabular(self, df_clean):
```

Comment:
it may be possible to speed this up using existing functionality
https://github.com/microsoft/recommenders/blob/71a38d422c00329f8f8226ea24ad6260b1b7b4e9/reco_utils/common/python_utils.py#L69

Reply:
Thank you for pointing this function out! I tested if it was faster to (1) keep the code as is, (2) use `python_utils.get_top_k_score_items()` in this method, or (3) reduce iterating through k and not use `python_utils.get_top_k_score_items()`.

k = 200
| Method | Time to execute `__organize_results_as_tabular()` |
|---|-------|
| (1) Original | 0.266 seconds |
| (2) Use `python_utils.get_top_k_score_items()` | 0.353 seconds |
| (3) Reduce iterating in original | 0.228 seconds |

This is just a rough comparison, but I did a few runs and took the mean (for when k=200) and saw that approach (3) was the fastest. My latest commit uses this approach.

Also, I noticed when playing around with this that `python_utils.get_top_k_score_items()` breaks when you set `sort_top_k=True`. If I were to try to use `python_utils.get_top_k_score_items()`, I would need to go through some extra steps to make sure the output is properly sorted.